test/kotlin/integration/CancelStreamTest.kt

```diff
@@ -0,0 +1,155 @@
+        address:
+          socket_address: { protocol: TCP, address: 0.0.0.0, port_value: 10000 }
+        api_listener:
+          api_listener:
+            "@type": $hcmType
+            stat_prefix: api_hcm
+            route_config:
+              name: api_router
+              virtual_hosts:
+              - name: api
+                domains: ["*"]
+                routes:
+                - match: { prefix: "/" }
+                  route: { cluster: fake_remote }
+            http_filters:
+            - name: envoy.filters.http.platform_bridge
+              typed_config:
+                "@type": $pbfType
+                platform_filter_name: $filterName
+            - name: envoy.router
+              typed_config:
+                "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+      clusters:
+      - name: fake_remote
+        connect_timeout: 0.25s
+        type: STATIC
+        lb_policy: ROUND_ROBIN
+        load_assignment:
+          cluster_name: fake_remote
+          endpoints:
+          - lb_endpoints:
+            - endpoint:
+                address:
+                  socket_address: { address: 127.0.0.1, port_value: 10101 }
+    """
+
+class CancelStreamTest {
+
+  init {
+    JniLibrary.loadTestLibrary()
+  }
+
+  private val filterExpectation = CountDownLatch(1)
+  private val runExpectation = CountDownLatch(1)
+
+  class CancelValidationFilter(
+    private val latch: CountDownLatch
+  ) : ResponseFilter {
+    override fun onResponseHeaders(headers: ResponseHeaders, endStream: Boolean): FilterHeadersStatus<ResponseHeaders> {
+      return FilterHeadersStatus.Continue(headers)
+    }
+
+    override fun onResponseData(body: ByteBuffer, endStream: Boolean): FilterDataStatus<ResponseHeaders> {
+      return FilterDataStatus.Continue(body)
+    }
+
+    override fun onResponseTrailers(trailers: ResponseTrailers): FilterTrailersStatus<ResponseHeaders, ResponseTrailers> {
+      return FilterTrailersStatus.Continue(trailers)
+    }
+
+    override fun onError(error: EnvoyError) {}
+
+    override fun onCancel() {
+      latch.countDown()
+    }
+  }
+
+  @Test
+  fun `cancel stream calls onCancel callback`() {
+    val engine = EngineBuilder(Custom(config))
+      .addPlatformFilter(
+        name = "cancel_validation_filter",
+        factory = { CancelValidationFilter(filterExpectation) }
+      )
+      .setOnEngineRunning {}
+      .build()
+
+    val client = engine.streamClient()
+
+    val requestHeaders = RequestHeadersBuilder(
+      method = RequestMethod.GET,
+      scheme = "https",
+      authority = "example.com",
+      path = "/test"
+    )
+      .addUpstreamHttpProtocol(UpstreamHttpProtocol.HTTP2)
+      .build()
+
+    client.newStreamPrototype()
+      .setOnCancel {
+        runExpectation.countDown()
+      }
+      .start(Executors.newSingleThreadExecutor())
+      .sendHeaders(requestHeaders, false)
+      .cancel()
+
+    engine.terminate()
```

Comment:
? How come we are terminating before the expectations are fulfilled?

Reply:
Let me change it to be after the expectation awaits